### PR TITLE
Simplify `fem::Expression`

### PR DESCRIPTION
**Pull request by garth-wells**
C++ Expression no longer requires a mesh at construction, which supports fixing of #2363 and will avoid some unnecessary JIT. The mesh is now passed to the `Expression::eval` function.

Also remove `dict()` default parameters, use `None` instead.

### CODE REVIEW

**Review by nate-sime — APPROVED** [review]
Looks great.

File: python/dolfinx/fem/function.py [hunk 1]
```diff
@@ -131,51 +146,55 @@ def __init__(self, ufl_expression: ufl.core.expr.Expr, X: np.ndarray,
...
-            )._cpp_object
+            self._argument_function_space = arguments[0].ufl_function_space()._cpp_object
         else:
-            raise RuntimeError(
-                "Expressions with more that one Argument not allowed.")
+            raise RuntimeError("Expressions with more that one Argument not allowed.")
 
         def create_expression(dtype):
-            if dtype is np.float32:
+            if dtype == np.float32:
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply python/dolfinx/fem/function.py:170]: @nate-sime I don’t quite follow, could you expand?

File: python/dolfinx/fem/function.py [hunk 2]
```diff
@@ -83,9 +84,9 @@ def __complex__(self):
 
 
 class Expression:
-    def __init__(self, ufl_expression: ufl.core.expr.Expr, X: np.ndarray,
-                 form_compiler_options: dict = {}, jit_options: dict = {},
-                 dtype=default_scalar_type):
+    def __init__(self, e: ufl.core.expr.Expr, X: np.ndarray,
+                 comm: typing.Optional[_MPI.Comm] = None, form_compiler_options: dict = {},
+                 jit_options: dict = {}, dtype=None):
```
Comment (jorgensd) [inline python/dolfinx/fem/function.py:89]: We should really use `jit_options: Optional[dict] = None`, ref https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python

File: python/dolfinx/fem/function.py [hunk 3]
```diff
@@ -659,12 +676,12 @@ def VectorFunctionSpace(mesh: Mesh,
...
                                   element.degree(), element.lagrange_variant,  # type: ignore
                                   element.dpc_variant, element.discontinuous,  # type: ignore
-                                  shape=(mesh.geometry.dim, ) if dim is None else (dim, ),
+                                  shape=(mesh.geometry.dim,) if dim is None else (dim, ),
                                   gdim=mesh.geometry.dim, rank=1)
     except AttributeError:
         ed = ElementMetaData(*element)
         ufl_e = basix.ufl.element(ed.family, mesh.basix_cell(), ed.degree,
-                                  shape=(mesh.geometry.dim, ) if dim is None else (dim, ),
+                                  shape=(mesh.geometry.dim,) if dim is None else (dim, ),
```
Comment (jorgensd) [inline python/dolfinx/fem/function.py:684]: Why this change in formatting?

Comment (garth-wells) [reply python/dolfinx/fem/function.py:684]: autopep8 formatter change.
